README.md
![execel logo](https://raw.githubusercontent.com/Descent098/ezspreadsheet/master/.github/logo.png)

# EZ Spreadsheet

*A simple API to store/load python objects to/from spreadsheets*

## Table of contents
- [Goals](#goals)
- [Installation](#installation)
- [Quick-start](#quick-start)
- [Additional Documentation](#additional-documentation)

## Goals
This project has a few goals:
1. Make OO projects easier to serialize to xlsx
2. Make use of Spreadsheet serialization and deserialization (currently xlsx and csv supported)
3. Write the simplest to use api for object serialization

## Installation

### From PyPi

1. Run ```pip install ezspreadsheet``` or ```sudo pip3 install ezspreadsheet```

### From source

1. Clone this repo: (https://github.com/Descent098/ezspreadsheet)
2. Run ```pip install .``` or ```sudo pip3 install .```in the root directory


## Quick-start

### Store some animal instances in a spreadsheet called 'animals.xlsx'
```python
from ezspreadsheet import Spreadsheet

class Animal():
    def __init__(self, name:str, conservation_status:str):
        self.name = name
        self.conservation_status = conservation_status

leopard_gecko = Animal('Leopard Gecko', 'Least Concern')

philippine_eagle = Animal('Philippine Eagle', 'Threatened')

with Spreadsheet('animals.xlsx', Animal) as output_sheet:
    output_sheet.store(leopard_gecko, philippine_eagle)
```

### Store a list of instances into a spreadsheet called 'users.xlsx'
```python
from ezspreadsheet import Spreadsheet

import random
import string
from dataclasses import dataclass

@dataclass
class User():
    Name:str
    Age:int
    Weight:int
    Family: list # Note that Iterables will be flattened to a string with newline seperators

instances = []
ranstring = lambda: ''.join(random.choices(string.ascii_uppercase, k=10)) # Generates a random 10 character string
for i in range(1000):
    instances.append(User(ranstring(), random.randint(12,100), random.randint(75,400), [ranstring(), ranstring(), ranstring()]))

with Spreadsheet('users.xlsx', User) as output_sheet:
    output_sheet.store(instances)
```

## Additional Documentation

Additional documentation can be found at https://kieranwood.ca/ezspreadsheet

For details on how contributing to the project, please see [CONTRIBUTING.md](https://github.com/Descent098/ezspreadsheet/blob/master/CONTRIBUTING.md), for details on upcoming changes see [our roadmap](https://github.com/Descent098/ezspreadsheet/projects).

For most recent changes see [CHANGELOG.md](https://github.com/Descent098/ezspreadsheet/blob/master/CHANGELOG.md).


CHANGELOG.md
# Changelog

## 0.2.2; November 24th 2020

- Updated links to include docs and source code urls

## 0.2.1; November 24th 2020

- Updated description to match name change

## 0.2.0; November 24th 2020

- Added deserialization methods for xlsx files
- Added serialization and deserialization methods for CSV files
- Changed name from ezexcel to ezspreadsheet
- Split Xlxs processing to internal class and converted Spreadsheet class to dispatching class

## 0.1.1; September 25th 2020

Fixed logo loading on PyPi

### Bug fixes

- Fixed loading issue with logo on PyPi

## 0.1.0; September 25th

Initial release of EZ Excel

### Features

- Ability to provide a class to instantiate a Spreadsheet
- Ability to pass instances in an iterable of class to Spreadsheet to be serialized
- Automatically flatten Iterable attributes within instances to endline delimited strings
- Added testing suite for all existing functionality

### Documentation improvements

- Wrote all existing documentation :)


tests/ezexcel_test.py
# Standard library dependencies
import os                           # Used to validate file existence, and remove test files
import random                       # Used to generate random values for test classes
import string                       # Used to filter random values for test classes
from unittest import mock           # Used to fake results for tests with inputs
from dataclasses import dataclass   # Used to save time instantiating test classes

# Internal Dependencies
from ezexcel import Spreadsheet     # Functionality being tested

# Third Party Dependencies
from openpyxl import load_workbook  # Used to load a workbook and validate values


# Create test classes

@dataclass
class User():
    # Test dataclass
    Name:str
    Age:int
    Weight:int
    Family: list # Note that Iterables will be flattened to a string with newline seperators


class Animal():
    # Test standard class
    def __init__(self, name:str, conservation_status:str):
        self.name = name
        self.conservation_status = conservation_status


def test_single_instance():
    """Validates that a single instance can be stored to a sheet"""

    # Setup test instance
    leopard_gecko = Animal('Leopard Gecko', 'Least Concern')

    # Write test instance to spreadsheet
    with Spreadsheet('animals.xlsx', Animal) as output_sheet:
        output_sheet.store(leopard_gecko)
    
    # Validate data saved properly
    workbook = load_workbook(filename = 'animals.xlsx', data_only=True)
    sheet = workbook.active
    # Confirm heading is correct
    assert sheet["A1"].value == "name"
    assert sheet["B1"].value == "conservation_status"

    # Check the A2 and B2 values are equal to leopard_gecko
    assert sheet["A2"].value == leopard_gecko.name
    assert sheet["B2"].value == leopard_gecko.conservation_status

    os.remove("animals.xlsx")  # remove animals.xlsx



def test_multi_instance():
    """Validates that multiple single instance can be stored to a sheet"""

    # Setup test instances
    leopard_gecko = Animal('Leopard Gecko', 'Least Concern')

    philippine_eagle = Animal('Philippine Eagle', 'Threatened')

    # Write test instances to a sheet
    with Spreadsheet('animals.xlsx', Animal) as output_sheet:
        output_sheet.store(leopard_gecko, philippine_eagle)

    # Validate data saved properly
    workbook = load_workbook(filename = 'animals.xlsx', data_only=True)
    sheet = workbook.active
    # Confirm heading is correct
    assert sheet["A1"].value == "name"
    assert sheet["B1"].value == "conservation_status"

    # Check the A2 and B2 values are equal to leopard_gecko
    assert sheet["A2"].value == leopard_gecko.name
    assert sheet["B2"].value == leopard_gecko.conservation_status
    
    # Check the A3 and B3 values are equal to philippine_eagle
    assert sheet["A3"].value == philippine_eagle.name
    assert sheet["B3"].value == philippine_eagle.conservation_status

    os.remove("animals.xlsx")  # remove animals.xlsx

def test_iterable_instances():
    """Validates that an iterable of instances can be stored to a sheet"""
    instances = []
    ranstring = lambda: ''.join(random.choices(string.ascii_uppercase, k=10)) # Generates a random 10 character string
    instances.append(User("John Doe", 20, 75, ["Abby", "Mike", "Janice"]))
    for i in range(1000):
        instances.append(User(ranstring(), random.randint(12,100), random.randint(75,400), [ranstring(), ranstring(), ranstring()]))

    with Spreadsheet('users.xlsx', User) as output_sheet:
        output_sheet.store(instances)

    # Validate data saved properly
    workbook = load_workbook(filename = 'users.xlsx', data_only=True)
    sheet = workbook.active
    # Confirm heading is correct
    assert sheet["A1"].value == "Name"
    assert sheet["B1"].value == "Age"
    assert sheet["C1"].value == "Weight"
    assert sheet["D1"].value == "Family"

    # Check the second row is equal to the first instance
    assert sheet["A2"].value == "John Doe"
    assert sheet["B2"].value == 20
    assert sheet["C2"].value == 75
    assert sheet["D2"].value == "Abby\nMike\nJanice\n"

    os.remove("users.xlsx")  # remove users.xlsx


def test_flattened_list_attributes():
    """Validates that list elements are flattened on save"""
    test_user = User("John Doe", 20, 75, ["Abby", "Mike", "Janice"])

    with Spreadsheet('users.xlsx', User) as output_sheet:
        output_sheet.store(test_user)

    # Validate data saved properly
    workbook = load_workbook(filename = 'users.xlsx', data_only=True)
    sheet = workbook.active

    # Check the list has been flattened
    assert sheet["D2"].value == "Abby\nMike\nJanice\n"

    os.remove("users.xlsx")  # remove users.xlsx


def test_unmatched_instances():
    """Validates that an error is thrown when mismatched instance types are tried to be stored"""
    # Setup test instances
    leopard_gecko = Animal('Leopard Gecko', 'Least Concern')

    philippine_eagle = Animal('Philippine Eagle', 'Threatened')

    # Try to use mismatching instances of class (User and Animal)
    try:
        with Spreadsheet('users.xlsx', User) as output_sheet:
            output_sheet.store(leopard_gecko, philippine_eagle)
        os.remove("users.xlsx")  # remove users.xlsx
        assert False # Test fails because it allows to store a different class than provided
    except ValueError:
        assert True


noxfile.py
import nox

def confirm(message:str, validators:list = ["y", "yes"]):
    """Used to confirm you have done what is being asked"""
    valid_input = False

    while not valid_input:
        response = input(message + str(validators).replace(","," or").replace("[","(").replace("]",")").replace("'",""))
        if response in validators:
            valid_input = True
        else:
            raise ValueError(f"Failed to confirm; {message}")

@nox.session
def build(session):
    """Builds a distribution for PyPi"""

    # Create source distribution
    session.run("python", "setup.py", "sdist")

    # Create wheelfile
    session.install("wheel")
    session.run("python", "setup.py", "bdist_wheel", "--universal")

@nox.session
def release(session):
    """Runs build and then releases distribution to PyPi"""
    build(session)

    # Confirm all the essential release stuff has been done
    confirm("Have you run the tests?")
    confirm("Have you updated inline docs?")
    confirm("Have you updated the docs folder documentation?")
    confirm("Have you created the release page?")
    confirm("Have you updated the readme docs?")
    
    session.install('twine')
    session.run("twine", "upload", "dist/*")

@nox.session(python=["3.6", "3.7", "3.8"])
def test(session):
    """Runs tests defined in tests folder against all installed versions
        of python from 3.6-3.8"""
    session.install('.')
    session.install('pytest')
    session.run('pytest')

@nox.session
def docs(session):
    # Serve documentation to verify it's how you want
    session.install(".") # install ezexcel
    session.install("pdoc3")
    session.run("pdoc3", "ezspreadsheet", "--http", "localhost:8080")


ezexcel.py
"""A simple class based xlsx serialization system

Limitations
-----------
Currently only classes with 51 or less attributes are supported

Examples
--------
### Store some animal instances in a spreadsheet called 'animals.xlsx'
```
from ezexcel import Spreadsheet

class Animal():
    def __init__(self, name:str, conservation_status:str):
        self.name = name
        self.conservation_status = conservation_status

leopard_gecko = Animal('Leopard Gecko', 'Least Concern')

philippine_eagle = Animal('Philippine Eagle', 'Threatened')

with Spreadsheet('animals.xlsx', Animal) as output_sheet:
    output_sheet.store(leopard_gecko, philippine_eagle)
```

### Store a list of instances into a spreadsheet called 'users.xlsx'
```
from ezexcel import Spreadsheet

import random
import string
from dataclasses import dataclass

@dataclass
class User():
    Name:str
    Age:int
    Weight:int
    Family: list # Note that Iterables will be flattened to a string with newline seperators

instances = []
ranstring = lambda: ''.join(random.choices(string.ascii_uppercase, k=10)) # Generates a random 10 character string
for i in range(1000):
    instances.append(User(ranstring(), random.randint(12,100), random.randint(75,400), [ranstring(), ranstring(), ranstring()]))

with Spreadsheet('users.xlsx', User) as output_sheet:
    output_sheet.store(instances)
```
"""
import logging                               # Used to log data for debugging
import datetime                              # Used to validate type assertions for datetime instances
from typing import Union, Iterable           # Used for type hinting and type assertions

import colored                               # Colours terminal output for emphasis
from openpyxl import Workbook                # Used to open and operate with xlsx files
from openpyxl.styles import Font, Alignment  # Used to pretty output to files


class Spreadsheet():
    """A class that takes in instances of objects and serializes them to xlsx files

    Parameters
    ----------
    file_name : (str)
        The name of the .xlsx file that will be saved out (extension can be included or excluded)

    class_identifier : (object)
        The class object for instances you want to store, see example(s) for details

    Raises
    ------
    ValueError

        In two cases:

            1. If instances provided to Spreadsheet.store() do not match type used to construct Spreadsheet instance
            2. If class provided has more than 51 attributes (see limitations section of docs for details)

    Examples
    --------
    #### Store some animal instances in a spreadsheet called 'animals.xlsx'
    ```
    from ezexcel import Spreadsheet

    class Animal():
        def __init__(self, name:str, conservation_status:str):
            self.name = name
            self.conservation_status = conservation_status

    leopard_gecko = Animal('Leopard Gecko', 'Least Concern')

    philippine_eagle = Animal('Philippine Eagle', 'Threatened')

    with Spreadsheet('animals.xlsx', Animal) as output_sheet:
        output_sheet.store(leopard_gecko, philippine_eagle)
    ```
    """
    def __init__(self, file_name:str, class_identifier:object):
        self.file_name = file_name
        self.class_identifier = class_identifier
        self.workbook = None
        self.worksheet = None

        # Make sure filename has .xlsx extension
        if not file_name.endswith(".xlsx"):
            logging.debug(f"added .xlsx to {file_name}")
            file_name += ".xlsx"

        # Get all attributes of class defined in __init__
        self.class_attributes = class_identifier.__init__.__code__.co_varnames[1::] # Skip the self
        if len(self.class_attributes) > 51:
            raise ValueError(f"Provided class {class_identifier.__name__} has more than 51 attributes")


    def __enter__(self):
        """Entrypoint for the context manager

        Returns
        -------
        Spreadsheet
            Reference to self
        """
        self.workbook = Workbook()
        self.worksheet = self.workbook.active
        self.worksheet.page_setup.fitToWidth = 1
        return self


    def __exit__(self, exc_type, exc_value, traceback):
        """Exitpoint for the context manager

        Returns
        -------
        bool
            True if the context manager ran into no issues saving files
        """
        if exc_type is None and exc_value is None:
            try:
                self.workbook.save(self.file_name)
                print(f"{self.file_name} successfully saved")
            except PermissionError:
                input(f"{colored.fg(1)}File {self.file_name} is currently open{colored.fg(15)}\nPlease close it and hit enter to save file: ")
                self.workbook.save(self.file_name)
            return True

        else:
            print(f"{colored.fg(1)}Ran into exception {exc_type}, with value {exc_value} here is traceback{colored.fg(15)}")
            return False


    def _add_row_to_spreadsheet(self, data:list, row:int, style:Font = False):
        """Take in some data, and an int for a row and add that data to that row

        Parameters
        ----------
        data : list
            The data you want to store in the row

        row : int
            The index of the row to store the data to

        style : Font, optional
            If you want to supply custom formatting for the row, by default False
        """
        # The value that will be converted using chr() for column identifiers i.e. A1 B1 etc.
        column_identifier = 65  # Initialize to ord() value of 'A'

        for value in data:
            if column_identifier == 91:  # Roll over to Ax column identifiers from x column identifiers
                label = f"AA{row}"
            elif column_identifier > 91:  # If beyond Z in column identifiers
                label = f"A{chr(column_identifier-26)}{row}"
            else:  # If before or at Z in column identifiers
                label = f"{chr(column_identifier)}{row}"
            logging.debug(f"{value} will be written to {label}")

            # Apply styles if specified
            if style:
                self.worksheet[label].font = style

            # Add value to worksheet
            if isinstance(value, Iterable) and not type(value) in [str, int, float]:
                # If value is an Iterable that's not a str, int or float then flatten it to a str
                flattened_value = ""
                for sub_value in value: 
                    flattened_value += f"{str(sub_value)}\n"
                self.worksheet[label] = flattened_value
            elif type(value) not in [str, int, float, datetime.datetime]:
                # Value is not a str, int, float or datetime object (all can be natively serialized)
                self.worksheet[label] = str(value)
            else: # If value is a string, int, float or datetime object
                self.worksheet[label] = value

            # Apply wrap text formatting to all rows that aren't the heading
            if not row == 1: 
                self.worksheet[label].alignment = Alignment(wrapText=True)

            # Increment the column identifiers variable to move to next column letter
            column_identifier += 1


    def _get_values_from_instance(self, instance:object) -> list:
        """Get's the instance's attribute values

        Parameters
        ----------
        instance : object
            The instance to pull the attribute values from

        Returns
        -------
        list
            The values for the attributes from the instance
        """
        logging.debug(f"Attributes are {self.class_attributes}")
        values = [] # All the values of the attributes in order
        for attribute in self.class_attributes:
            logging.debug(f"Looking for attribute {attribute} found value {instance.__dict__[attribute]}")
            values.append(instance.__dict__[attribute]) 
        return values


    def store(self, *instances:Union[object, Iterable[object]]):
        """Takes in instance(s) of the specified class to store

        Parameters
        ----------
        instances : (Iterable[object] or arbitrary number of isntances)
            The instances with the data you want to store

        Raises
        ------
        ValueError
            If an instance is not the correct type
        
        Examples
        --------
        #### Store some animal instances in a spreadsheet called 'animals.xlsx'
        ```
        from ezexcel import Spreadsheet

        class Animal():
            def __init__(self, name:str, conservation_status:str):
                self.name = name
                self.conservation_status = conservation_status

        leopard_gecko = Animal('Leopard Gecko', 'Least Concern')

        philippine_eagle = Animal('Philippine Eagle', 'Threatened')

        with Spreadsheet('animals.xlsx', Animal) as output_sheet:
            output_sheet.store(leopard_gecko, philippine_eagle)
        ```
        """
        print(f"Beginning to store {self.class_identifier.__name__} instances to {self.file_name}")
        current_row = 1  # The current row that the iteration is at

        # Add heading with the list of class attributes to A1
        self._add_row_to_spreadsheet(self.class_attributes, current_row, Font(bold=True, size=14))
        current_row += 1  # Increment row to start with row right after heading
        logging.debug(f"Instances are {instances}")

        # Check if instance provided is a class of correct type, or an Iterable
        for current_instance in instances:
            logging.debug(f"Instance is {str(current_instance)}")
            if isinstance(current_instance, Iterable):  # If argument is an Iterable (i.e. list, tuple etc.)
                for sub_instance in current_instance:
                    if not isinstance(sub_instance, self.class_identifier):  # Validate sub-instance is correct type
                        raise ValueError(f"Provided instance: {sub_instance} is not of type {self.class_identifier}")
                    else:
                        self._add_row_to_spreadsheet(self._get_values_from_instance(sub_instance), current_row)
                        current_row += 1
            elif not isinstance(current_instance, self.class_identifier):  # If argument is not correct type
                raise ValueError(f"Provided instance: {current_instance} is not of type {self.class_identifier}")
            
            else:  # If argument is a single class instance of the correct type
                logging.debug(f"Adding values from {str(current_instance)}: {self._get_values_from_instance(current_instance)}")
                self._add_row_to_spreadsheet(self._get_values_from_instance(current_instance), current_row)
                current_row += 1
